### ISSUE TITLE
Support multi-branch, multi-channel, LTS and pre-release workflows

### ISSUE THREAD

**Issue by pvdlg** [semantic-release/semantic-release#563]
This a proposal to support more complex release workflows in a fully automated way, including:
- LTS branches and distribution channels
- Future branches and distribution channels
- Pre-releases

Please see https://github.com/semantic-release/evolution/pull/1 for the proposed list of workflows to be supported by `semantic-release`.

**`semantic-release` modifications**
- [ ] Add the new `branches` options
- [ ] Normalize each branch configuration based on [naming convention](https://gist.github.com/pvdlg/9fa33dbc6792edc2cf3d0765c5d8b8b3#branches-naming-convention)
- [ ] [Validate](https://gist.github.com/pvdlg/9fa33dbc6792edc2cf3d0765c5d8b8b3#configuration-validation) the `branches` configuration
- [ ] Re-purpose the `branch` options as the current branch on which `semantic-release` is running
- [ ] Verify that only allowed release are made
  - [ ] On LTS branches verify it's in `range`
  - [ ] In default and future branches make sure that releasing will comply with the `type` of the following non pre-release future branch
- [ ] When no version has to be released, check if we are in a merge situation (the commit associated with the last release is present in following branch  and it is not a pre-release branch)
  - [ ] If yes, make a release in order for the publish plugins to update the channel
- [ ] Determine if the branch is a pre-release one, and if it is increment the version accordingly
- [ ] Pass the `channel` on which to release `nextRelease` to the publish plugins
- [ ] Move the git tag creation to the core

**`publish` plugins modifications**
- [ ] If release already exists, make it available on `nextRelease.channel` (in case of github, remove the `prerelease` flag if the target channel is the default one)
- [ ] For the `npm` plugin, always publish on the dist-tag passed in `nextRelease.channel` or `latest` if `nextRelease.channel` is not set
- [ ] For the `github` plugin, set `prerelease` flag if the current branch is a future or prerelease
- [ ] Remove the tag creation from the `publish` plugins

**Deprecate `getLastRelease` plugins**
- [ ] Move the `getLastRelease` logic of the git plugin into the core

This is fairly broad and complex subject, so I'd love to have feedback from the community, in particular:
- Is there a type of workflow that wouldn't be supported with this new feature?
- Is there some potential inconsistencies that are not covered? (i.e. would it be possible to create a release that shouldn't happen, or create an incoherent situation).
- Would this feature create issue with a specific publishing platform? Or should this feature be adapted to support a specific platform via plugin?

**Comment by gr2m** [comment]
Wow, you put lots of thoughts into this. I love the idea of "Branches naming convention"!

Here some notes from my first review. I’ll give it some more thoughts with suggestions on how to move forward :)

> Distribution channels to make unstable version available to early adopters. Usually named `next` or `unstable`.

I’ve never seen a dist tag called `unstable`. Maybe just leave it at "Usually named `next`?

> A software meant to be used by final users

end users

> "channel"

I wonder if instead of "channel" we should call it "npm-dist-tag", to make it super clear what it is?

> push to 1.x.x or 2.x.x branches

I think it’s more common to call them `1.x` and `2.x` branches

> (Pre-releases) "branches": ["master", "4.0.0-beta", "5.0.0-alpha"]

When I push to the `4.0.0-beta` branch and it releases a `4.0.0-beta` version to npm, what happens with the GitHub release notes? I think there would be a conflict with the branch `4.0.0-beta` as the release would need a git tag with the same name, I _think_ that is not possible, but I haven’t checked it.

I’m also not sure if releasing these pre versions to `@latest` is the right approach? Because if I do `npm install my-pkg` I get `my-pakg@4.0.0-beta` if it’s released to latest, I’m not sure if that is the intent? Shouldn’t these rathe be relesed to `@beta` release channel instead?

**Comment by pvdlg** [comment]
> I’ve never seen a dist tag called unstable. Maybe just leave it at "Usually named next?

I think `unstable` is kind of the naming convention in the Linux world, mostly because Debian name future releases like that. But mentioning only `next` probably make more sense.

> I wonder if instead of "channel" we should call it "npm-dist-tag", to make it super clear what it is?

The objective was to use a generic name that is not attached to a specific package manager like `npm`, as we want to support any type of release target, not only `npm`. Channel is the generic term used by Google for this type of distribution: https://support.google.com/chromebook/answer/1086915?hl=en

> I think it’s more common to call them 1.x and 2.x branches

Both works, as long it's a valid semver range it will be detected as a LTS branch. I'll make it more clear in the doc.

> When I push to the 4.0.0-beta branch and it releases a 4.0.0-beta version to npm, what happens with the GitHub release notes? I think there would be a conflict with the branch 4.0.0-beta as the release would need a git tag with the same name, I think that is not possible, but I haven’t checked it.

The tag will be 4.0.0-beta.X where X is the iteration: 4.0.0-beta.0, 4.0.0-beta.1, 4.0.0-beta.2 etc... But that's a good feedback, I'll make sure we'll never end up with tag/branch conflict.

> I’m also not sure if releasing these pre versions to @latest is the right approach? Because if I do npm install my-pkg I get my-pakg@4.0.0-beta if it’s released to latest, I’m not sure if that is the intent? Shouldn’t these rathe be relesed to @beta release channel instead?

As far as I understand if you do `npm install my-pkg` you will not get my-pakg@4.0.0-beta, you will get the highest non pre-release version available on `@latest`. I'll test to make sure.

**Comment by felixfbecker** [comment]
> As far as I understand if you do npm install my-pkg you will not get my-pakg@4.0.0-beta, you will get the highest non pre-release version available on @latest. I'll test to make sure.

Afaik `npm install my-pkg` will always get you `latest`, and if that was set to a prerelease, then you will get a prerelease. I already opened many issues because package maintainers forgot to publish with `--dist-tag=next` and accidentally tagged `latest` to a prerelease. But maybe npm changed it?

**Comment by schickling** [comment]
This is highly appreciated. Our current workflow is:

1. Create a new branch (e.g. `dev`)

2. Assuming you're on the `dev` branch, you need to add the following to your package.json to publish using the `next` NPM tag:

```json
  "release": {
    "branch": "dev"
  },
  "publishConfig": {
    "tag": "next"
  }
```

3. Before merging back to `master` (typically via PR) remove everything from (2) again.

@pvdlg do you (or someone else) have suggestions on how to improve this workflow until this issue has been implemented and released?

**Comment by pvdlg** [comment]
@schickling thanks for the feedback!

Actually I don't really understand how to workflow you are describing could currently works. Let's say you have a version `1.0.0` released on npm which correspond to the sate of the `master` branch. The `1.0.0` has been made available on the `@latest` dist-tag.
Then you create a `dev` branch (step 1), set the config (step 2) and push a feature commit on it. That will trigger a `1.1.0` release that will be available only on the `@next` dist-tag. Then you change the config and merge to `master` (step 3). That will trigger a release process from `master` that will attempt to release `1.1.0` and make it available on `@latest`, but that would fail as `1.1.0` already exists.

Can you clarify how you make that works? Is what you are trying to achieve is to make the code that only on `dev` available to only early adopters?

**Comment by schickling** [comment]
Ideally we'd be able to "promote" released versions by assigning the `latest` (or another) tag.
So far we've been releasing versions like `1.1.0-beta.1`, `1.1.0-beta.2` etc. How would we do something like this using `semantic-release`?

**Comment by pvdlg** [comment]
Ah ok, I think I understand better now. The workflow is:
- push on `dev` => release a beta version available only `@next`
- push on master (by merging from `dev`) => release a normal version available on `@latest` dist-tag

That's not currently possible as we don't have a way to make pre-release (i.e. `1.1.0-beta.1`) in semantic-release currently. The new feature would allow to do that with a [pre-release branch](https://github.com/semantic-release/evolution/blob/release-workflows/_drafts/release-workflows.md#pre-releases).

What you can do now is something similar to what we do in the semantic-release repo:
- Only one branch `master` (named `caribou` for this particular repo) 
- Push to master => release a normal version available on `@next` dist-tag
- When we want to promote a release (make it available for everyone) we make it available on `@latest` with `npm dist-tag add semantic-release@<version to promote> latest `

**Comment by priley86** [comment]
guys... please do a better job of announcing these breaking changes in a README or something... i love this tool and i rarely complain about others open source work, but this is quickly becoming a mysterious library that I'm somewhat afraid of upgrading. Please do your best to ensure ensure error messages are helpful for the end user.

For anyone else upgrading from an earlier version and searching these github issues, you may receive an error like this:

```
> semantic-release pre && npm publish && semantic-release post
  Usage: semantic-release [options]
  Run automated package publishing
  Options:
    -b, --branch <branch>                Branch to release from
    -r, --repositoryUrl <repositoryUrl>  Git repository URL
    --verify-conditions <paths>          Comma separated list of paths or packages name for the verifyConditions plugin(s)
    --get-last-release <path>            Path or package name for the getLastRelease plugin
    --analyze-commits <path>             Path or package name for the analyzeCommits plugin
    --verify-release <paths>             Comma separated list of paths or packages name for the verifyRelease plugin(s)
    --generate-notes <path>              Path or package name for the generateNotes plugin
    --publish <paths>                    Comma separated list of paths or packages name for the publish plugin(s)
    --debug                              Output debugging information
    -d, --dry-run                        Dry-run mode, skipping verifyConditions, publishing and release, printing next version and release notes
    -h, --help                           output usage information
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! patternfly-react@0.0.0-semantically-released semantic-release: `semantic-release pre && npm publish && semantic-release post`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the patternfly-react@0.0.0-semantically-released semantic-release script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.npm/_logs/2018-01-08T19_59_14_238Z-debug.log
```

This likely means you missed @pvdlg 's comment here with the suggested fix:
https://github.com/semantic-release/semantic-release/issues/411#issuecomment-353233730

With no errors/readme updates, i'm just really hoping someone else finds this issue since the original is now locked!

**Comment by felixfbecker** [comment]
@priley86 documentation of breaking changes belong to the release notes, not the readme. They were documented in the [release notes of 9.0.0](https://github.com/semantic-release/semantic-release/releases/tag/v9.0.0) with a migration guide. But a smart way to detect the `pre` argument may have been nice.

Let's not get this issue thread off topic though please.

**Comment by priley86** [comment]
Sure, thanks for the prompt response and I’ll be sure to check your release notes going forward.  I’m a big fan of this library and just hope upgrades continue to be fluid.

**Comment by n1ru4l** [comment]
Pretty excited about this 😊 This allows powerful workflows!

**Comment by travi** [comment]
a thought that has been on my mind around this type of flow for a while is how to support multiple pre-releases at the same time. i think npm itself is limited in this regard, but rapid releases when using `semantic-release` highlight my question even further. i'm happy to take this to another issue if it is too off-topic, but i'm curious around your thoughts about how this would work with this proposal.

scenario:

* i have a feature branch with breaking changes that i expect to bump the major version once merged to master. the changes have a big impact, so i want to publish and test a few `-alpha.#` or `-beta.#` releases and test them to ensure there are no surprises.
* a team member is working on a separate feature that also includes breaking changes that would bump the major version

how do i handle the unknown around the order of those breaking changes landing? truthfully, i dont know the actual version until the point of publishing (which is a big part of the point of semantic-release). so, i dont want my feature branch that was branched from `4.x.x` to result in `5.x.x-beta.1` because it might end up being a pre-release for `6.0.0` or even `7.0.0`, etc by the time it gets published as latest.

ideally, id like the pre-release to be more along the lines of `featureName-beta.1`, but i don't think that would be supported by npm, correct? what recommendations do you have and how do you think semantic-release can help with this?

**Comment by pvdlg** [comment]
In your example the order of the 2 breaking changes would dependents on what you merge and when and how you configure your branches.

If you have a `master` and a `4.0.0-beta` branches and the last release done from `master` is `3.1.1`:
- Pushing a breaking change on `4.0.0-beta` would release `4.0.0-beta.1` on the `beta` dist-tag
- Pushing another breaking change on `4.0.0-beta` would release `4.0.0-beta.2` on the `beta` dist-tag
- Merging `4.0.0-beta` would release `4.0.0` including the 2 breaking change on the `latest` dist-tag

If you have a `master` and `4.0.0-beta` branches and the last release done from `master` is `3.1.1`:
- Pushing a breaking change on `4.0.0-beta` would release `4.0.0-beta.1` on the `beta` dist-tag
- Merging `4.0.0-beta` would release `4.0.0` including the breaking change  on the `latest` dist-tag
- Pushing another breaking change on `4.0.0-beta` would be invalid as each branch need to have non overlapping range (so you can't have `master` and `4.0.0-beta` having the same range)

The idea of `pre-release` is that it doesn't respect semver, per semver specification :)
Whatever you push on the pre-release branch only the last digit increase (`4.0.0-beta.1`, `4.0.0-beta.2`, `4.0.0-beta.3` etc).
That allow to "group" breaking changes without making a major release. Once you merge to master, the not semver rules applies: breaking change => major release and as it is the case currently multiple breaking released at once create only one major increase.

That also mean that a `pre-release` branch has a lifetime limited to the development cycle of a new version (in that case `4.0.0`).
As a side note the "future branch" works differently as they defined for a release type and not a release version. So for example a branch defined as:
```
{
  "branch": "next",
  "type": "major
}
```
Will accept release range that depends on the releases existing on the previous branch in the config.
If the previous branch is `master` and the last release there is `3.1.1` then `next` will accept only commits that trigger a release in the range `4.x.x`. If the `type` was `minor` then it would accept only commits that trigger a release in the range `3.x.x`. If the `type` was `minor` then it would accept only commits that trigger a release in the range `3.1.x`.

"Future branches" are long-lived in the sense their configuration will never have to be changed.
The `type` property indicated what type of changes are allowed there.
The "pre-release branches", as mentioned before, live from the beginning to the end of the dev cycle for a given version.

**Comment by travi** [comment]
i think what i'm understanding from your clarification is that my scenario would follow the idea of a "future branch" rather than a "pre-release branch", which is a good distinction. as you mentioned, "pre-release" seems to be more for grouping changes into a single version bump.

if we focus on the idea of future branches in my scenario above, i wouldnt really have long lived branches like `next`, but more feature specific branches. if both features branched from `master` and the last release there is `3.1.1`. if a breaking change was introduced to the first branch, what would the first part of the version be? (<what goes here?>-beta.1). if a breaking change was then introduced to the second branch, same question...<what goes here?>-beta.1? in both cases, i hope there is an option other than `4.0.0`, but im not clear on if that part still needs to be semver compatible or not. ideally, id like that to simply be a string, possibly matching the branch name.

i'm not sure i fully understand the limitation based on `type`. is that an upper bound or a strict limit? even for branches that would end up in a major bump, i would still want to allow some commits that dont include breaking changes. would those be prevented?

**Comment by pvdlg** [comment]
The format `4.0.0-beta` is only for pre-release branch. That's the format defined by [semver](https://semver.org/#spec-item-9).
What I called "Future branch" is what is described in semantic-release/evolution#1.

I'm not sure I understand your example especially regarding feature branches. A feature (like a branch on whci your develop a feature and open a PR) will never trigger a release. The release will happens when you merge your PR to a release branch.
Basically if you push to master or if you merge a PR into master it's the same thing for semantic-release.

> i'm not sure i fully understand the limitation based on type. is that an upper bound or a strict limit? even for branches that would end up in a major bump, i would still want to allow some commits that dont include breaking changes. would those be prevented?

The `type` defines the type of release allowed on this branch. That allow to do something like:
- `master` => only `fix` and `feat`
- `next` => `fix`, `feat` and breaking change

That allow to do a workflow like that: https://github.com/sass/sass/blob/stable/CONTRIBUTING.md#branches

**Comment by travi** [comment]
> I'm not sure I understand your example especially regarding feature branches. A feature (like a branch on whci your develop a feature and open a PR) will never trigger a release. The release will happens when you merge your PR to a release branch.

maybe a little more context would help. what i'm referring to as a feature branch in this case would be a branch in the primary repo, not from a fork. i would want the alpha/beta/etc style release to be opt-in and not for every branch. most of my feature branches have a short life and dont need a pre-release before being merged into master and getting released as `latest`, but there are times where these branches are more complex and live long enough or are disruptive enough to benefit from a pre-release. 

two cases come to mind where this could happen at the same time as another feature branch and without disciplined planning to know which will be merged first. 
* the first is on my team when multiple team members are working on independent features. this case is a bit more random and less predictable because there are multiple people making changes without impacting each other. 
* the second is for side-projects that i dont find enough time to focus on getting features released that involve more complex changes. these often make a little bit of progress here and there when i can sit down and get a partial thought out of my head and into code. admittedly, pre-releases being difficult ends up further demotivating progress because the level of effort ends up being so much higher than typical auto-publishes from `master`

does that help clarify? these may not be as common as some of the other use cases, but i find them useful. if there is a way to support this approach, i think it could be pretty valuable to some users like me.

**Comment by pvdlg** [comment]
Just to be sure we are using the same vocabulary:
- pre-release: release version formatted as specified by SemVer (`x.y.z-preid.N`, for example `4.0.0-beta.1`)
- Future release: Release done on a dist-tag. Follow the normal SemVer (`x.y.z`)

> i would want the alpha/beta/etc style release to be opt-in and not for every branch.

Everything is always opt-in. semantic-release runs only on branches that are part of the config.

> most of my feature branches have a short life and dont need a pre-release before being merged into master and getting released as latest

If you don't add them to the semantic-release config this is exactly what's going to happen.

> but there are times where these branches are more complex and live long enough or are disruptive enough to benefit from a pre-release.

For those ones you will have to add them to the config as a "Pre-release branch"

Regarding the two example you gave, I think they are fairly common and they should be support. If you want to create a branch from which you want to do pre-releases you will have to add to the semantic-release config (as semantic-release needs to know you want to make releases from there).

If you want to make pre-releases you have to give semantic-release the version from which to start (like `4.0.0`) because I don't see a way to determine that automatically. Also in a SemVer pre-release the version part is mandatory (`4.0.0-beta.1` is valid while `beta.1` is not).

So in order to do what you mention in your examples:
1. Create a branch `side-project` from `master`
2. Update the semantic-release config to add the `side-project` branch and set the `prerelease` property `4.0.0-side-project`
3. Every `fix`, `feat`, `perf` or `BREAKING CHANGE` on `side-project` will trigger a pre-release (`4.0.0-side-project.1`, then `4.0.0-side-project.2`, then `4.0.0-side-project.3` etc... )
4. Once you merge to `master` the normal process will happen, each commit is evaluated and a release is done based on the last release (if the last release is `1.0.0` then we'll increase it based on the commit added, so it might be `1.1.0` or `2.0.0`)

As you can see in that example, the fact the pre-release branch specifies `4.0.0` is irrelevant and appear to be useless. But it's required by the SemVer spec, and you be able to publish on npm if the version number isn't valid. Unfortunately I don't think there is a way for semantic-release to guess or make up a version number. It also has the advantage to communicate intent to your users, for example `4.0.0-beta.1` should contains all the code released in `3.x.x` (semantic-release will check that btw).

**Comment by travi** [comment]
thanks. i think we've clarified what i was after. sorry if it created a bunch of noise to get here, but it helps me to have a more solid understanding around what can be possible. 

to restate what we've clarified around this:
* npm requires a version to be included when a package is published
* npm _only_ supports valid semver formatted versions, which does allow for pre-releases like `4.0.0-beta.1`, etc
* `feature-beta.1` and similar do not fall into this category because `feature` does not follow the `#.#.#` format
* what i would ideally like for supporting the scenarios that i described is not possible
* since npm does not support it, semantic-release cannot either

so with all of that said, if a project does manage to be disciplined enough to stay focused on limiting to intentional breaking changes that would bump from `3,x.x` to `4.0.0` to a single pre-release branch, what happens if `4.0.0-beta.1` and `4.0.0-beta.2` have already been released in that branch but a breaking change slips through to master before that pre-release is ready for promotion to `latest`? is the next pre-release then `5.0.0-beta.1`? is that limited by the branch name or other config? these are the confusion points i've always had about semver prereleases because you dont truly know what the version number will be until it gets fully released...

**Comment by pvdlg** [comment]
> sorry if it created a bunch of noise to get here

I don't think it's noise at all. It really helps to figure out the in and outs of this feature and all the scenario to test. This is exactly what this issue is made for. So no worries.

> so with all of that said, if a project does manage to be disciplined enough to stay focused on limiting to intentional breaking changes that would bump from 3,x.x to 4.0.0 to a single pre-release branch, what happens if 4.0.0-beta.1 and 4.0.0-beta.2 have already been released in that branch but a breaking change slips through to master before that pre-release is ready for promotion to latest?

The breaking change on `master` will trigger a `4.0.0` release.
The fact that `master` now have commits ahead of `side-project` will prevent you to merge `side-project` into `master` (that's git doing that). You will have to merge `master` into `side-project`.
Then on `side-project` each commit will continue to trigger versions like `4.0.0-beta.3`, `4.0.0-beta.4` etc... Here semantic-release could errors out and tell you that `side-project` is behind `master`, but I'm not really sure it's a good, as you might legitimate case in which you want to do that.

What's tricky here is that pre-release version live in their own "continuity", they don't respect the semver rules of regular version. Basically in pre-releases the version part is purely indicative and no rules (by semver or semantic-release) are applied on it.

To put in other terms, with regular releases, the version `3.1.1` guarantees the it contains all the code from all the release with a version number inferior. Btw, even when you revert a commit, the versioning move forward not backward, so in terms of releasing a revert is considered like a regular commit and it appears in the commit history log (it just happens to make the exact opposite change of another commit).

With pre-release version, nothing of that true. Or to be more precise nothing of that is enforced. You can release any version with any type of commit inside. They are made unique by using the last digit but that's it. It might be clever to release pre-releases with a number higher that the last stable release but it's not enforced because `4.0.0-beta.1` => major release => `4.0.0-beta.2` (and not `5.0.0-beta.1`).

Basically pre-release branches are used in that situation:
- I want to make a release on every push (so a non-release branch is not an option)
- But I don't want to increment versions normally because I'm going to do a bunch of breaking change and I don't want to end up with version 9999999999.0.0 (so a "Future branch" which increment versions normally and  releases on a dist-tag is not an options)

**Comment by travi** [comment]
> With pre-release version, nothing of that true. Or to be more precise nothing of that is enforced. You can release any version with any type of commit inside.

i guess that makes sense, but is a bit unfortunate. it definitely leaves the opportunity for the version to confuse a consumer rather than inform them of the risk of updating. i understand its a bit different in the case of pre-releases but it just feels like there should be a more clear way. i dont know what could be though when npm doesnt allow deferring the version choice. as you mentioned, there might be some valid cases for this, so maybe its not actually a bad thing

> What's tricky here is that pre-release version live in their own "continuity"

this is the part that i keep coming back to. i agree that they should live on their own until merged to master, but the requirement to decide the `#.#.#` when starting the series of pre-releases limits that ability in my mind. still a shortcoming of semver itself, not of this proposal

> the fact the pre-release branch specifies 4.0.0 is irrelevant and appear to be useless. But it's required by the SemVer spec

the idea of including the version number but ignoring it is at least a way around the limitation so that the feature name could be treated as the meaningful part. i suppose `alpha` or `beta` could still be included as well to communicate the stability.

> * pre-release: release version formatted as specified by SemVer (x.y.z-preid.N, for example 4.0.0-beta.1)
> * Future release: Release done on a dist-tag. Follow the normal SemVer (x.y.z)

this conversation has been very helpful in getting my head around semantic-release/evolution#1. the point above maybe points out something i've been overlooking. you mention that a future release is done on a dist-tag. would a pre-release be published to the `latest` channel? i would not have expected that if thats the case. i would only expect it to hit `latest` after it is no longer a pre-release.

**Comment by pvdlg** [comment]
> this is the part that i keep coming back to. i agree that they should live on their own until merged to master, but the requirement to decide the #.#.# when starting the series of pre-releases limits that ability in my mind. still a shortcoming of semver itself, not of this proposal

This is why with this proposal you have two way to handle release that you don't want to be available to everyone:
- Future branch, that have normal semver and release to a dist-tag. Version number is increased normally.
- Pre-release branches that use a pre-release semver and also release to a dist-tag. Version number is not increased normally (only the last digit).

If you want to increase the `x` in `x.y.z` when you push a breaking change, then you should use a "Future branch" not a "Pre-release" one.

> the idea of including the version number but ignoring it is at least a way around the limitation so that the feature name could be treated as the meaningful part. i suppose alpha or beta could still be included as well to communicate the stability.

To be valid a pre-release semver be formatted like `<version-<preid>.N>`. All the 3 parts are mandatory. In your example the `preid` is `beta` or `alpha` and it's mandatory.
The version is completely ignored. It doesn't participate in any semantic-release rule, but it's still there as its the one published (and for which a git tag is created)

> you mention that a future release is done on a dist-tag. would a pre-release be published to the latest channel? i would not have expected that if thats the case. i would only expect it to hit latest after it is no longer a pre-release.

Both are done in a dist-tag. In case of a pre-release, if the dist-tag is not specified the `preid` (i.e. beta in `4.0.0-beta`) will be used.

**Comment by pvdlg** [comment]
> if i have a pre-release of 4.0.0-some-feature-alpha.7 and want to promote it to 4.0.0-some-feature-beta.1 what steps would i use? since semantic release ignores the version and preid, i realize there wouldnt be provided process, but what would the manual steps be? is it as simple as renaming the branch?

To support this workflow you would have the branches `4.0.0-some-feature-alpha` and `4.0.0-some-feature-beta`. Once you merge `4.0.0-some-feature-alpha` into `4.0.0-some-feature-beta`, semantic-release will run on `4.0.0-some-feature-beta` and make the release ``4.0.0-some-feature-beta.<last_release + 1>`.

> if i have two future branches in addition to master with activity, do publishes on each impact the others since the version is incremented normally for each? i see that stalled branches are expected to be updated with changes from master, but the reverse can't be true if the changes aren't ready for promotion. is this type of a scenario simply a reason not to use future branches when these things are possible?

With "Future branches" you have a lot more rules that are verified, especially each one need to:
- Each one has to define a `type`, which defined the release gap allowed with  the previous branch
- The version range allowed on Future branches cannot overlap

For example you can define something like:
- `master` => only patch release on dist-tag `latest`
- `features` => only patch and minor releases on dist-tag `latest`
- `major` => patch, minor, major releases on dist-tag `next`

Pushing a commit on one branch doesn't have a direct impact other than what git already does: for merging `features` into `master`, `features` must have all the `master` commits.
So if you push a `fix` to `master` and you already have a few `feats` on `features`, in order to merge `features` into `master` you need to update `features` with the `fix` commit.

Normally that should avoid inconsistent situations (by verifying the releases are consistent with the git history).

**Comment by gr2m** [comment]
@travi when you have ongoing branches where you develop features, why do you need to publish it to npm at all? Why not install from git directly in these cases? Setting a version to "travi/repo-name#feature-branch-name" is simple enough? You could create `"travi/repo-name#some-feature-alpha.1` and if a breaking change comes up just create another branch and ask folks to give it a try by installing `"travi/repo-name#some-feature-alpha.2". I don’t see the benefit of npm here, people wouldn’t get new versions automatically with npm, but they would with the install from git, you’d have full control.

If you have feedback to  semantic-release/evolution#1 I’d suggest you comment directly on it, that’s why we do it :)

**Comment by travi** [comment]
> why do you need to publish it to npm at all? Why not install from git directly in these cases?

i used to do that type of thing quite a bit. once i started adding a build step to distribute common-js and es-module bundles, that isn't realistic anymore. `main` and `module` are defined so the bundles are loaded from `lib/`, which is ignored from git because that is the output directory of the build step. i dont know of a way to make this work when depending directly from git w/o adding pretty significant complexity to enabling the build process after installation.

in reality, we currently avoid dist-tags altogether at this point. this mostly works well because we are very disciplined to follow true continuous integration. there are exceptions where a pre-release would be very helpful for a risky change, so if semantic-release made the process easy to publish and iterate on pre-releases, we'd be far more likely to do it. we would still need to manually specify the version in the consumer, but it would give us simpler ability to test a version in a real environment before promoting it to latest.

semantic-release was transformative for our willingness to pull code out of applications to separate packages, so making this process easy has the potential to have a big impact as well.

> people wouldn’t get new versions automatically with npm

i'd love to see greenkeeper change that [if a project already depends on a pre-release](https://github.com/greenkeeperio/greenkeeper/issues/521) :)  

> If you have feedback to semantic-release/evolution#1 I’d suggest you comment directly on it, that’s why we do it :)

will do :)

**Comment by gr2m** [comment]
> i'd love to see greenkeeper change that if a project already depends on a pre-release :)

I mean that if you release a fix then people won’t get it automatically.  Greenkeeper might eventually be able to notify you about the changes, but just running `npm install` will not install the newest version because there is no semver version range that covers `2.0.0-beta.1` and `2.0.0-beta.2`, they would be treated like breaking changes.

**Comment by travi** [comment]
> just running npm install will not install the newest version because there is no semver version range that covers 2.0.0-beta.1 and 2.0.0-beta.2

i see your point, but that doesnt really impact me or my team, we dont really rely on semver ranges but instead use lockfiles and depend on greenkeeper to keep us up to date. because of https://github.com/greenkeeperio/greenkeeper/issues/506, we also have to pin exact versions of our direct dependencies, which has worked well for us so far. the only reason i am very interested in defining ranges would be to [let greenkeeper know the upper bound of my desired range](https://github.com/greenkeeperio/greenkeeper/issues/434), to prevent PRs for versions that I know my project isn't ready for yet.

either way, i'm still not sure how to handle the loss of the published artifacts that arent versioned with the source code

**Comment by micha149** [comment]
I am not sure if the auto-branch option is to tightly coupled to a specific workflow. You assume, that there is a branch for pre releases, but this might not be true. For example on a new project one would like to create alpha, beta, or whatever releases on the master branch until the project reaches a state where it makes sense to be released on a stable version.

What about simply adding an option to force the level which will be increased. `semver` also provides an `--preid` option which is used to control what goes after the dash. So I could put simply control the release names from my CI system.

Example:
```yaml
release-alpha-beta:
    stage: release
    script:
        - semantic-release --force-level prerelease --preid $CI_COMMIT_REF_NAME # this variable contains the branch name on gitlab
    only:
        - alpha
        - beta

release-dev:
    stage: release
    script:
        - semantic-release --force-level prerelease --preid pre
    only:
        - develop

release-master:
    stage: release
    script:
        - semantic-release
    only:
        - master
```

**Comment by domwashburn** [comment]
Hi there, I hate to post a non-productive comment, but I'm wondering when we might be able start working with this new tagging system. I'm reaching a point in my process where I can either work out a temporary solution until this is ready or wait on this and prioritize some other work.

I'm so glad I found this thread, I'm eagerly awaiting this update 🤗

**Comment by pvdlg** [comment]
The code is 99% completed and can be found in the [multi-branches](https://github.com/semantic-release/semantic-release/tree/multi-branches) branch. You should be able to start experimenting with it.
I have to finish writing the doc though which takes quite some time...

**Comment by axe312ger** [comment]
Actually, we'd love to use this at [Contentful](https://github.com/contentful), too :)

**Comment by Martaver** [comment]
Ditto over here

**Comment by axe312ger** [comment]
I know we should use the subscribe feature for `+1` and similar. Just wanted to point out that a release to any npm tag besides `latest` would be great. That way we can test the new code :)

**Comment by pvdlg** [comment]
The branch is badly outdated and I have to rebase it.
A PR will be opened and the code published to dist-tag when it's ready enough to test. I don't know when that will be.

**Comment by SimonFinney** [comment]
@pvdlg Same thought as the above - I can dedicate time to help get this brought into the mainline

**Comment by mattyclarkson** [comment]
@pvdlg happy to help out also

**Comment by pvdlg** [comment]
Fixed by #991

**Comment by pvdlg** [comment]
This feature as been implemented in version `16.0.0-beta.X` and released on the `@beta` dist-tag.

So far we didn't had any bug reports on `16.0.0-beta.X`. Before moving it to the `@next` dist-tag, and later `@latest`, that would be great to have some feedback from real-life usage.

If you have tested or used `16.0.0-beta.X` and made successful releases, please add a comment to this thread so we can get an idea of how much it was tested and feel confident moving forward with the general release. Please add any details you see fit (plugins used, type of releases, link to the repo etc...).

If you experienced an issue, please open an issue on the GitHub tracker, so we can fix it while we are still in beta.

If you have questions or other feedback, please start a thread  on https://spectrum.chat/semantic-release.

Thanks!

**Comment by jfairley** [comment]
I'm interested in giving this a go on one of my team's projects. 

I'm curious about one thing though, and rather than try it in the wild: Based on the N.N.X branch name strategy, if I set up a branch like 1.2.x, and if someone were to mistakenly commit a `feat`, would I end up with a 1.2.1 (based on the branch name pattern) or a 1.3.0 (based on the latest commit)?

**Comment by bddckr** [comment]
Besides https://github.com/semantic-release/github/issues/139 I have no issue with the latest betas and that may just be a result of me accidentally not using the _latest_ beta versions. This issue isn't really a deal-breaker anyway.

<details><summary>Versions used (Click to expand)</summary>

```
        "semantic-release": "^16.0.0-beta.8",
        "@semantic-release/commit-analyzer": "^7.0.0-beta.1",
        "@semantic-release/release-notes-generator": "^7.1.4",
        "@semantic-release/changelog": "^3.0.1",
        "@semantic-release/git": "^7.1.0-beta.1",
        "@semantic-release/npm": "^5.2.0-beta.3",
        "@semantic-release/github": "^5.3.0-beta.3"
```

</details>

<details><summary>Configuration (Click to expand)</summary>

```
    "release": {
        "branches": [
            "+([1-9])?(.{+([1-9]),x}).x",
            "latest",
            "master",
            "next",
            "next-major",
            {
                "name": "prerelease",
                "prerelease": "prerelease"
            },
            {
                "name": "beta",
                "prerelease": "beta"
            },
            {
                "name": "alpha",
                "prerelease": "alpha"
            }
        ],
        "plugins": [
            [
                "@semantic-release/commit-analyzer",
                {
                    "parserOpts": {
                        "revertPattern": "/^(?:Revert|revert:)\\s\"?([\\s\\S]+?)\"?\\s*This reverts commit (\\w*)\\./"
                    },
                    "releaseRules": "./release-rules.js"
                }
            ],
            [
                "@semantic-release/release-notes-generator",
                {
                    "parserOpts": {
                        "revertPattern": "/^(?:Revert|revert:)\\s\"?([\\s\\S]+?)\"?\\s*This reverts commit (\\w*)\\./"
                    }
                }
            ],
            [
                "@semantic-release/changelog",
                {
                    "changelogFile": "./CHANGELOG.md"
                }
            ],
            [
                "@semantic-release/npm",
                {
                    "tarballDir": "./npm"
                }
            ],
            [
                "@semantic-release/git",
                {
                    "message": "chore(release): set version to ${nextRelease.version}\n\n${nextRelease.notes}",
                    "assets": [
                        "../../../CHANGELOG.md",
                        "../../Package/package.json"
                    ]
                }
            ],
            [
                "@semantic-release/github",
                {
                    "assets": [
                        {
                            "path": "./npm",
                            "label": "Unity Package"
                        }
                    ]
                }
            ]
        ]
    }
```

</details>

**Comment by bniedermeyer** [comment]
I've had some issues with publishing to channels with a private registry, but haven't tried with the latest betas yet. I'll try again with those this week and open something up if I'm still running into problems.

**Comment by appsolutegeek** [comment]
Hi, I have tried testing it, but I get a problem saying

```
[10:00:56 AM] [semantic-release] › ℹ  This test run was triggered on the branch 1.16.x, while semantic-release is configured to only publish from master, therefore a new version won’t be published.
```

I have .releaserc.json file created like so - I added the .json extension so it was easier to manged in my IDE which according to the docs - is fine.

```
{
  "branches": [
    ",+([1-9])?(.{+([1-9]),x}).x",
    "master",
    "next",
    "next-major",
    {
      "name": "beta",
      "prerelease": true
    },
    {
      "name": "alpha",
      "prerelease": true
    }
  ],
  "plugins": [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",
    "@semantic-release/changelog",
    [
      "@semantic-release/npm",
      {
      }
    ],
    "@semantic-release/git"
  ]
}
```
 
And I have just created a branch from version 1.16 tag and branch is called 1.16.x but it doesn't seem to recognise it.

```
➜  semantec-release-poc git:(1.16.x) npx semantic-release --no-ci
[10:09:23 AM] [semantic-release] › ℹ  Running semantic-release version 16.0.0-beta.8
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/changelog"
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/npm"
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/git"
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/release-notes-generator"
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/changelog"
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/npm"
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/git"
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/npm"
[10:09:23 AM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/npm"
[10:09:30 AM] [semantic-release] › ℹ  This test run was triggered on the branch 1.16.x, while semantic-release is configured to only publish from master, therefore a new version won’t be published.
```

**Comment by appsolutegeek** [comment]
my dev dependencies are 

```
  "devDependencies": {
    "semantic-release": "^16.0.0-beta.8",
    "@semantic-release/commit-analyzer": "^7.0.0-beta.1",
    "@semantic-release/release-notes-generator": "^7.1.4",
    "@semantic-release/changelog": "^3.0.1",
    "@semantic-release/git": "^7.1.0-beta.1",
    "@semantic-release/npm": "^5.2.0-beta.3",
    "@semantic-release/github": "^5.3.0-beta.3"
  }
```

**Comment by pvdlg** [comment]
@appsolutegeek I think the problem is the branch `1.16.x` doesn't exists on your remote repo.
For several reasons we have to get the list of branches in your repo from `origin`. So this is expected.

If you still encounter the issue while having the branch on `origin`, please open a new issue.

**Comment by pvdlg** [comment]
@gr2m the issue with the `@latest` in the comment has been fixed in [v16.0.0-beta.4](https://github.com/semantic-release/semantic-release/releases/tag/v16.0.0-beta.4%40beta).

The issue didn't received a comment because it was opened when semantic-release ran: https://travis-ci.org/octokit/rest.js/jobs/467622755#L765.

This is something on GitHub side. When a PR has a `fix #xx` in the description and get merged it should close the referenced issue. I noticed recently that often GitHub does not close the issue.

**Comment by appsolutegeek** [comment]
> @appsolutegeek I think the problem is the branch `1.16.x` doesn't exists on your remote repo.
> For several reasons we have to get the list of branches in your repo from `origin`. So this is expected.
> 
> If you still encounter the issue while having the branch on `origin`, please open a new issue.

Thanks, yes that was it! Thanks
I have a small issue remaining but I will open a new issue as suggested

Thanks

**Comment by bddckr** [comment]
I can confirm pre-release as well as regular releases work without issues now. I'm using the npm plugin without publishing and then release a new GitHub release with that tarball using the GitHub plugin.

**Comment by bniedermeyer** [comment]
After running with the latest betas I was unable to fix my issues. I've opened up #1040.

**Comment by pvdlg** [comment]
@gr2m, re https://github.com/semantic-release/semantic-release/issues/563#issuecomment-447679389

It turns out GitHub does not close issues solved by a PR unless this PR is merged to the default branch. With https://github.com/semantic-release/github/pull/146 we will now comment in this case.

I think GitHub should close the issue when the PR get merged to any protected branch, rather than just the default one. Worst case a GitHub app can do that.

**Comment by travi** [comment]
i just finished up a prerelease where i created an `alpha` branch, (`alpha` this time, but `beta` is typically be more common for my flow. i expect that it would behave the same, though), pushed changes as normal and `semantic-release` worked very nicely to publish multiple `...-alpha.1`, `...-alpha.2`, etc versions to the `alpha` dist tag. once i merged the branch, the new version was published to the `latest` dist tag as expected.

this project just uses the default config with no config file and i publish from travis using `npx semantic-release@beta`

well done @pvdlg!

**Comment by travi** [comment]
also, i really like how the github release notes for the full release after the prereleases rolls all of the changes up from the prereleases. i think that'll be really handy for consumers that don't subscribe to non-latest channels

**Comment by gr2m** [comment]
I just set this up for [nock](https://github.com/nock/nock). I successfully released a beta pre-release https://github.com/nock/nock/releases/tag/v11.0.0-beta.1%40beta, but after I merged more pull requests into "beta" the publish failed with the error

> **semantic-release** cannot push the version tag to the branch `undefined` on remote Git repository with URL `https://[secure]@github.com/nock/nock.git`.

https://github.com/nock/nock/issues/1304. [build error on travis](https://travis-ci.org/nock/nock/jobs/469667109#L499)

I think the problem might have been that I merged multiple pull requests into quick succession, before the current build could finish. The [last build](https://travis-ci.org/nock/nock/jobs/469669423#L481) was able to finish and did not have the problem

**Comment by pvdlg** [comment]
The `undefined` in the message should be solved in https://github.com/semantic-release/semantic-release/pull/1043.

 However you have had the message `The local branch beta is behind the remote one, therefore a new version won't be published.` and the build should have not failed. I don't understand why it didn't work...

That would be helpful to have debug messages. I think that would be a great idea to always enable `--debug` or `DEBUG=semantic-release:*` for anyone testing the beta.

**Comment by SimonFinney** [comment]
I can confirm this is working for me with the following configuration -

```js
{
  branches: [
    'master',
    {
      name: 'dev',
      prerelease: 'prerelease',
      channel: 'next',
    },
  ]
}
```

Thanks @pvdlg!

**Comment by gr2m** [comment]
When the `git push --dry-run https://[secure]@github.com/nock/nock.git HEAD:beta` command fails due to [`(non-fast-forward)` error](https://travis-ci.org/nock/nock/jobs/470173796#L486), it creates an issues saying that the problem is "The push permission to the Git repository is required."

https://github.com/nock/nock/issues/1327

nock is using 16.0.0-beta.13

**Comment by pvdlg** [comment]
It's the same issue as https://github.com/semantic-release/semantic-release/issues/563#issuecomment-448403202. When this error occurs we call [`isBranchUpToDate`](https://github.com/semantic-release/semantic-release/blob/beta/index.js#L77) to check if it's due the local branch being behind or to another problem.
The error `! [rejected]        HEAD -> beta (non-fast-forward)` indicate the local branch is behind the remote one, so `isBranchUpToDate` should return `false` and semantic-release should exit with `false`. However it seems `isBranchUpToDate` returns `true` in that case but I don't know why.

Having the debug logs would be helpful.

**Comment by renemr86** [comment]
First of all, awesome tool! We've been using semantic-release for some time on Java projects with Gradle and CircleCI and it has been great!

Now we decided to move on to the beta version and publish releases on multiple branches. Until now, everything is working like a charm.

Thanks

**Comment by renemr86** [comment]
Hello

I'm trying to replicate the example [Working on a third future release](https://github.com/semantic-release/semantic-release/blob/beta/docs/recipes/pre-releases.md#working-on-a-third-future-release), however when I commit on alpha branch the semantic-release set the next release version with the next beta version instead of the next alpha version.

Example repository:
https://github.com/renemr86/semantic-release-test

.releaserc:
```
tagFormat: ${version}

branches:
  - master
  - name: next
    prerelease: 'rc'
  - name: beta
    prerelease: 'beta'
  - name: alpha
    prerelease: 'alpha'

verifyConditions:
  - "@semantic-release/git"
generateNotes: []
prepare: []
publish: []
success: []
fail: []
```

```
$ npx semantic-release --no-ci
[12:55:00 PM] [semantic-release] › ℹ  Running semantic-release version 16.0.0-beta.15
[12:55:00 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/git"
[12:55:00 PM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[12:55:00 PM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/npm"
[12:55:00 PM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/github"
[12:55:15 PM] [semantic-release] › ✔  Run automated release from branch alpha
[12:55:21 PM] [semantic-release] › ✔  Allowed to push to the Git repository
[12:55:21 PM] [semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/git"
[12:55:21 PM] [semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/git"
[12:55:21 PM] [semantic-release] › ℹ  Found git tag 2.0.0-beta.2@beta associated with version 2.0.0-beta.2 on branch alpha
[12:55:21 PM] [semantic-release] › ℹ  Found 1 commits since last release
[12:55:21 PM] [semantic-release] › ℹ  Start step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[12:55:21 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: feat: first feature of other release

BREAKING CHANGE: it breaks something
[12:55:21 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The release type for the commit is major
[12:55:21 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analysis of 1 commits complete: major release
[12:55:21 PM] [semantic-release] › ✔  Completed step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[12:55:21 PM] [semantic-release] › ℹ  The next release version is 2.0.0-beta.3
[12:55:24 PM] [semantic-release] › ✔  Created tag 2.0.0-beta.3@alpha
[12:55:24 PM] [semantic-release] › ✔  Published release 2.0.0-beta.3
```

```
$ git log --graph --oneline --all --decorate 
* 461efd6 (HEAD -> alpha, tag: 2.0.0-beta.3@alpha, origin/alpha) feat: first feature of other release
* bbbaf5f (tag: 2.0.0-beta.2@beta, origin/beta, beta) feat: second feature
* 4409cd5 (tag: 2.0.0-beta.1@beta) feat: first feature
| * 6bca0e4 (tag: 1.0.1, origin/master, master) fix: a fix
|/  
* cc1021a (tag: 1.0.0) feat: initial commit
```

**Comment by travi** [comment]
i just published a couple of beta versions. first a `fix`, then a `BREAKING CHANGE`. the first [published a minor pre-release](https://github.com/travi-org/matt.travi.org-components/releases/tag/v2.9.1-beta.1%40beta), as expected. for the second, i expected the pre-release to be bumped to a major, but a [second pre-release of the minor was published](https://github.com/travi-org/matt.travi.org-components/releases/tag/v2.9.1-beta.2%40beta).

is this the expected incrementing behavior? from the log, it appears that it was [correctly detected as a major change](https://travis-ci.org/travi-org/matt.travi.org-components/builds/482718295#L765).

**Comment by bddckr** [comment]
AFAIK this is expected - a pre-release **always** seems to advance by just updating the pre-release number, the "actual version" number always stays the same.

I tried finding an explanation of this in the [documentation](https://github.com/semantic-release/semantic-release/blob/beta/docs/recipes/pre-releases.md) but I don't think it clearly states this behavior. In case it really doesn't (I may have missed it) it should not just explicitly note the behavior for breaking changes, but instead note that this is true for **any** form of commit that would normally change the version.

**Comment by pvdlg** [comment]
Yes that's the intended behavior. That allows you when working on a pre-release to make as many fix, feature or breaking change without incrementing the version.
Once you merge the pre-release branch to a regular branch then normal version will be released.

**Comment by travi** [comment]
gotcha. im familiar with the behavior once things get merged, but i guess i had an incorrect assumption about the pre-release versioning. i assumed it would increment the patch pre-release until a more impactful (minor/major) change was added and then would publish a minor/major pre-release instead. fair enough if thats not the expectation

**Comment by felixfbecker** [comment]
I agree that it's unintuitive - if we know that the next non-pre release will have a breaking change, shouldn't the prerelease be for the next major version? I.e. if at least one breaking change is detected on a prerelease branch, the major should be bumped (once). More breaking changes would then only bump the prerelease number.

**Comment by travi** [comment]
i've been thinking about this some more and thought it could be helpful to at least share a bit more about where my assumptions/hopes came from. Not necessarily trying to convince anyone about a change, but at least maybe adding a few more thoughts around the topic:

> a pre-release always seems to advance by just updating the pre-release number, the "actual version" number always stays the same

i mostly agree with this and have followed that mentality when manually publishing pre-releases. however, i see two problems that i've either run into in that approach that i think could be different when using semantic-release. 

manually, i could think through the changes that i expect to make and do the semantic math myself to set the initial "version" to match the most impactful change i expect to make. if i know i'll be making a breaking change, ill make the inital pre-release suggest that. with semantic-release, it depends on which change is made first. in my case above, i made a commit that triggered only a patch before making a commit that would normally trigger a major. admittedly, this time that order was to test out what semantic-release would do, but i've followed that approach legitimately many times in the past (and was why i was curious).

the second piece is that i could simply be wrong. i may not realize that a series of pre-releases that i expected to only add new functionality would end up forcing me to break a public api. in normal use of semantic-release, i purposely dont fret about these types of things because i know it will handle translation of my changes into the proper semantic version.

in addition, theres always the possibility that master releases an independent change that increments the version published to latest beyond what the pre-release is referencing. that ends up being downright confusing for the next pre-release, so the ideal behavior, in my mind, would be that the pre-release version would account for that as well after merging master into the pre-release branch.

while i agree that the version part of the pre-release tends to be somewhat meaningless, i think semantic-release has enough information available to potentially make it more meaningful. as a consumer, it is certainly helpful to understand, from the semantic version, what the impact of the change is. when pulling in a pre-release, that information is a bit different, since it is the impact relative to what has been deployed as latest. however, without the updates that i'm referring to, its more relative to what _was_ deployed as latest when the first pre-release was cut and only suggests the impact of the first change(s) that resulted in the publish of that first pre-release.

as i said, i can see both sides and am not necessarily trying to convince that there should be a change, but hopefully these details help articulate what my assumptions were since i think there is a potential opportunity for semantic-release to provide further semantic value for pre-releases.

**Comment by renemr86** [comment]
I'm using semantic-release on a pre-release branch (beta). When I commit a break change, the semantic-release correctly indicates that a new major version will be generated, however instead of increasing the major version (4.0.0-beta.1), it keeps the current version (3.38.0) and only increases the pre-release version (3.38.0-beta.15 to  3.38.0-beta.16).

```
npx semantic-release
[3:27:51 PM] [semantic-release] › ℹ  Running semantic-release version 16.0.0-beta.18
[3:27:51 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/git"
[3:27:51 PM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[3:27:51 PM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/exec"
[3:27:51 PM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/exec"
[3:27:51 PM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/exec"
[3:27:51 PM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/npm"
[3:27:51 PM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/github"
[3:27:51 PM] [semantic-release] › ⚠  This run was not triggered in a known CI environment, running in dry-run mode.
[3:28:11 PM] [semantic-release] › ⚠  Run automated release from branch beta in dry-run mode
[3:28:16 PM] [semantic-release] › ✔  Allowed to push to the Git repository
[3:28:16 PM] [semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/git"
[3:28:16 PM] [semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/git"
[3:28:16 PM] [semantic-release] › ℹ  Found git tag 3.38.0-beta.15@beta associated with version 3.38.0-beta.15 on branch beta
[3:28:16 PM] [semantic-release] › ℹ  Found 2 commits since last release
[3:28:16 PM] [semantic-release] › ℹ  Start step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[3:28:16 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: ci(circle/semantic-release): upgrade semantic-release
[3:28:16 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The commit should not trigger a release
[3:28:16 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: feat(something): description

BREAKING CHANGE: description
[3:28:16 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The release type for the commit is major
[3:28:16 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analysis of 2 commits complete: major release
[3:28:16 PM] [semantic-release] › ✔  Completed step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[3:28:16 PM] [semantic-release] › ℹ  The next release version is 3.38.0-beta.16
[3:28:16 PM] [semantic-release] › ⚠  Skip step "prepare" of plugin "@semantic-release/exec" in dry-run mode
[3:28:16 PM] [semantic-release] › ⚠  Skip 3.38.0-beta.16@beta tag creation in dry-run mode
[3:28:16 PM] [semantic-release] › ⚠  Skip step "publish" of plugin "@semantic-release/exec" in dry-run mode
[3:28:16 PM] [semantic-release] › ⚠  Skip step "publish" of plugin "@semantic-release/exec" in dry-run mode
[3:28:16 PM] [semantic-release] › ✔  Published release 3.38.0-beta.16
[3:28:16 PM] [semantic-release] › ℹ  Release note for version 3.38.0-beta.16:
```

**Comment by gr2m** [comment]
I just tried to release a `7.x` release for Probot. I think everything was setup created, the `7.x` branch was created based on the last 7x release tag (`v7.5`). But instead of creating `7.5.1`, it created a tag `v7.0.0@7.x` and I’m not sure what exactly it did on npm?

Here are the full logs: https://travis-ci.org/probot/probot/jobs/501226804

**Comment by gr2m** [comment]
Oh it actually got rogue and after "releasing" 7.0.0, it continued with all the other existing tags :( https://travis-ci.org/probot/probot/jobs/501226804#L595 we canceled it eventually

**Comment by rodoabad** [comment]
How do you use a pre-release version? `npm i semantic-release@next`?

**Comment by sanderaernouts** [comment]
@pvdlg I'm trying this with `16.0.0-beta.22` but I'm getting this message when running it on branch `alpha`:
```
[10:32:36 PM] [semantic-release] » i  This test run was triggered on the branch alpha, while semantic-release is configured to only publish from master, therefore a new version won’t be published.
```

**Comment by haldunanil** [comment]
@pvdlg does the pre-release workflow support blobs in branch names? for example, I want to have a branch named `release/v2.0.0`, where the `vX.0.0` changes as I go up the branch names and each iteration is deleted after being merged into `master`. Can I configure `semantic-release` to recognize branch names by using a blob like `release/*` and then have all of them tagged with `beta`?

**Comment by paulfelton** [comment]
Any estimate on when this is likely to be merged?

**Comment by travi** [comment]
see https://github.com/semantic-release/semantic-release/issues/1134
